openSearchLane.go

```diff
@@ -0,0 +1,277 @@
+
+	osl2 := openSearchLane{
+		openSearchConnection: osl.openSearchConnection,
+	}
+	osl2.wg.Add(1)
+	atomic.AddInt32(&osl.openSearchConnection.refCount, 1)
+
+	ll.clone(&osl2.logLane)
+	osl2.logLane.writer = log.New(&osl2, "", 0)
+
+	return &osl2
+}
+
+func (osl *openSearchLane) DeriveReplaceContext(ctx context.Context) Lane {
+	ll := deriveLogLane(&osl.logLane, context.WithValue(ctx, parent_lane_id, osl.LaneId()), osl.tees, osl.cr)
+
+	osl2 := openSearchLane{
+		openSearchConnection: osl.openSearchConnection,
+	}
+	osl2.wg.Add(1)
+	atomic.AddInt32(&osl.openSearchConnection.refCount, 1)
+
+	ll.clone(&osl2.logLane)
+	osl2.logLane.writer = log.New(&osl2, "", 0)
+
+	return &osl2
+}
+
+func newOpenSearchClient(openSearchUrl, openSearchPort, openSearchUser, openSearchPass string) (client *opensearchapi.Client, err error) {
+	client, err = opensearchapi.NewClient(
+		opensearchapi.Config{
+			Client: opensearch.Config{
+				Transport: &http.Transport{
+					TLSClientConfig: &tls.Config{InsecureSkipVerify: true},
+				},
+				Addresses: []string{fmt.Sprintf("%s:%s", openSearchUrl, openSearchPort)},
+				Username:  openSearchUser,
+				Password:  openSearchPass,
+			},
+		},
+	)
+
+	return
+}
+
+func (osl *openSearchLane) Write(p []byte) (n int, err error) {
+	osl.openSearchConnection.mu.Lock()
+	defer osl.openSearchConnection.mu.Unlock()
+
+	logEntry := string(p)
+
+	fmt.Println(logEntry)
+
+	parentLaneId, ok := osl.Context.Value(parent_lane_id).(string)
+	if !ok {
+		parentLaneId = ""
+	}
+
+	logData := &LogTemplate{
+		AppName:      osl.openSearchConnection.appName,
+		ParentLaneId: parentLaneId,
+		JourneyID:    osl.journeyId,
+		LaneID:       osl.LaneId(),
+		LogMessage:   logEntry,
+	}
+
+	osl.openSearchConnection.logBuffer = append(osl.openSearchConnection.logBuffer, logData)
+
+	return len(p), nil
+}
+
+func (osl *openSearchLane) flushLogBuffer() {
+	backoffDuration := 10 * time.Second
+	ticker := time.NewTicker(backoffDuration)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-ticker.C:
+			osl.openSearchConnection.mu.Lock()
+			if len(osl.openSearchConnection.logBuffer) > 0 {
+				err := osl.bulkInsert(osl.openSearchConnection.logBuffer)
+				if err != nil {
+					backoffDuration *= 2
+					if backoffDuration > 10*time.Minute {
+						backoffDuration = 10 * time.Minute
+					}
+					ticker.Reset(backoffDuration)
+				} else {
+					backoffDuration = 10 * time.Second
+					osl.openSearchConnection.logBuffer = osl.openSearchConnection.logBuffer[:0]
+					ticker.Reset(backoffDuration)
+				}
+			}
+			osl.openSearchConnection.mu.Unlock()
+		case wg := <-osl.openSearchConnection.refCountCh:
+			osl.openSearchConnection.mu.Lock()
+
+			refCounter := atomic.LoadInt32(&osl.openSearchConnection.refCount)
+			fmt.Println(refCounter)
+
+			if wg != nil {
+				wg.Done()
+			}
+
+			if refCounter == 0 {
+				fmt.Println("hmmm")
+				if len(osl.openSearchConnection.logBuffer) > 0 {
```

Comment:
Presumably you'd make `func (osc *openSearchConnection) flush()` that does the business of sending the logBuffer to rest api.

Then in this `select`, the regular case calls `osc.flush()`, and so does this code when refCounter is zero.

Reply:
done